### ISSUE TITLE
Lasloader - 'Cannot enlarge memory arrays' on many repeat loads

### ISSUE THREAD

**Issue by oomcreative** [visgl/loaders.gl#1026]
I'm working with the lasLoader with Lidar data and it's great. But for the project I'm needing to load multiple Lidar datasets in a sequence, and then reload them one by one every couple of minutes (it's for a visual display), but after about 10 or so loads the loader gives this error.

`las-loader.worker.js:1 Cannot enlarge memory arrays. Either (1) compile with -s TOTAL_MEMORY=X with X higher than the current value 117440512, (2) compile with ALLOW_MEMORY_GROWTH which adjusts the size at runtime but prevents some optimizations, or (3) set Module.TOTAL_MEMORY before the program runs.`

To achieve this error I simply iterated the loader 10 or so times on the same file within a browser context. Is there a way to avoid this memory issue (by reseting/clearing the loader between loads) or advice on how best to implement either 1,2, or 3 to enable endless loads without causing memory issue.

**Comment by ibgreen** [comment]
@oomcreative Thanks for reporting. Haven't seen this error before, but this is an interesting use case. 

The problem seems to be repeated use of the worker. You could get around it by not using workers (`parse(..., LASLoader, {worker: false})` but then you would of course be blocking the main thread while you parse.

The LASLoader is special in that it is based on a chunk of code compiled from C to JavaScript so it is not easy to debug. 

A quicker way might be to offer an option to not cache workers, but restart them every time. The hit should be relatively small as the worker script is still loaded and cached by the browser.

Your thoughts?

**Comment by oomcreative** [comment]
thanks @ibgreen  I did another test of each time post-load simply refreshing the browser and reloading etc - this avoided the issue, which aligns with what you've said above. I think it would be best not to block the thread - so not caching the workers would be a good option. The use case here is from an onsite installation display - so no concerns regards download times etc, all runs locally. However to ensure it runs in offline context have had to use local workerUrl for the worker. Not sure if that has had any impact on this.

```
las: {
      workerUrl: process.env.PUBLIC_URL + "/lasmodule/las-loader.worker.js",
    },
```
So how would one not cache the workers? is this in the lasloader.worker.js script, or an option in las-loader? not sure how to implement this approach

**Comment by ibgreen** [comment]
The changes would start somewhere here: https://github.com/visgl/loaders.gl/blob/master/modules/loader-utils/src/lib/worker-utils/worker-farm.js#L21

We could for instance support this with `maxConcurrency: false` or some other option.

Take a look. If this seems like something you want to attempt yourself, I can provide more guidance.

**Comment by oomcreative** [comment]
@ibgreen thanks for the advice on this. Was hoping to find a solution without needing to rebuild the loader - is that possible? - by putting in a parameter into the request? trying to limit any custom build steps into the project - but if there is a way to force the worker to reload each time that would be great to know how

**Comment by ibgreen** [comment]
Yes we may need to modify loaders.gl to offer the ability to restart loaders.

We could try to set `maxConcurrency: 0` option but not sure if that would work, will have to look through the code.

**Comment by oomcreative** [comment]
Looks like this is being look into - that's great. not sure how to test/check this out, but let me know if it can be solved

**Comment by dryabinin-actionengine** [comment]
Good day. Fix was merged. To try this out you should use *reuseWorkers*
flag:
Docs are here:

   -
   https://github.com/visgl/loaders.gl/blob/master/docs/developer-guide/using-worker-loaders.md
   -
   https://github.com/visgl/loaders.gl/blob/master/modules/core/docs/api-reference/parse.md#options

**Comment by ibgreen** [comment]
@dryabinin-actionengine  We should publish a new version, please work with @vdunaev.

**Comment by vdunaev** [comment]
Hi everyone! I've published 2.3.7 patch with this fix

**Comment by oomcreative** [comment]
wow, this approach works perfectly. Have just installed 2.3.7 and added '{worker: true, reuseWorkers: false}' - and simulated  40+ sequential loads of lidar data using LASloader, and no 'Cannot enlarge memory arrays' anymore. thanks everyone for working through this
